opengovsg/postmangovsg
# Release v1.8.1

## Improvements
- Extract templating function into a shared module that can be used by both frontend and backend
- Log jobs using id instead of recipient, so as to be able to send messages with duplicate recipients
- Enable developers to use localstack in place of an s3 bucket, cloudwatch logs, secrets manager
- Enable manual override of a halted campaign - set halted to null to prevent the lambda from stopping the campaign
- Include error code in downloadable report
- Support email callbacks from Sendgrid

## Bug fixes
- Remove slashes in amplify.yml which caused Amplify console to crash
- Halt only on hard bounce
- Only allow camelcase violations for input and output vars

## Deploy notes
- Deploy /v1/email on API Gateway
- Set appropriate CALLBACK_SECRET in Travis
- Set appropriate callback url in Sendgrid
- Modify subscriptions on SNS to call the lambda via https endpoint.
- `ALTER TABLE campaigns ALTER COLUMN halted DROP NOT NULL;`


## Tests
- Templating should work for valid recipient, valid param
- Templating should fail for empty params, non-alphanumeric params
- Templating should pass for no params
- Send two messages with the same recipient - both should be updated by separate callbacks
- Send message using SES - SNS. Callback should occur.
- Send message using SES - HTTP. Callback should occur.
- Send message using Sendgrid - HTTP. Callback should occur.
- Halt a campaign, retry it by setting halted to false. Campaign should halt again
- Halt a campaign, retry it by setting halted to null. Campaign should not halt again
- Download a report
